### PR TITLE
fix: typing issues with xgboost

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

Include a `FrozenDict` implementation for freezing default dictionary for better typing

### CODE REVIEW

File: bentoml/_internal/types.py [hunk 1]
```diff
@@ -268,6 +268,23 @@ def _format_dt(d: t.Union[datetime, str], _: type) -> datetime:
 cattr.register_structure_hook(datetime, _format_dt)  # type: ignore[misc]
 
 
+class FrozenDict(t.Dict[str, t.Any]):
```
Comment (bojiang) [inline bentoml/_internal/types.py:271]: This solution actually breaks the consensus of typing. We devs see the arg type is Dict, they will assume that they can do `dict.update` on it.

File: bentoml/_internal/frameworks/xgboost.py [hunk 2]
```diff
@@ -78,9 +79,9 @@ def _get_model_info(
 @inject
 def load(
     tag: t.Union[str, Tag],
-    booster_params: t.Optional[t.Dict[str, t.Union[str, int]]] = None,
+    booster_params: t.Dict[str, t.Union[str, int]] = FrozenDict(),
```
Comment (bojiang) [inline bentoml/_internal/frameworks/xgboost.py:82]: This solution actually breaks the consensus of typing. We devs see an arg which is Dict, they will assume that they can do `dict.update` on it.

File: bentoml/_internal/frameworks/xgboost.py [hunk 3]
```diff
@@ -116,10 +119,10 @@ def load(
 @inject
 def save(
     name: str,
-    model: "xgb.core.Booster",
+    model: "xgb.Booster",
     *,
-    booster_params: t.Optional[t.Dict[str, t.Union[str, int]]] = None,
-    metadata: t.Optional[t.Dict[str, t.Any]] = None,
+    booster_params: t.Dict[str, t.Union[str, int]] = FrozenDict(),
```
Comment (parano) [inline bentoml/_internal/frameworks/xgboost.py:124]: FrozenDict will break `_setup_booster_params` below, having default value for non primitive types is generally not recommended

File: bentoml/_internal/frameworks/xgboost.py (hunk 2 above)
Comment (aarnphm) [reply bentoml/_internal/frameworks/xgboost.py:82]: but the idea of FrozenDict is that `update` or all ops is forbidden if the given parameters is not given?

**Comment by aarnphm** [comment]
chatted offline with Bojiang.

**Pull request closed without merging**